File: opensource/test/src/main/java/com/sos/scheduler/engine/test/TestConfigurationBuilder.java
package com.sos.scheduler.engine.test;

import com.google.common.collect.ImmutableMap;
import com.sos.scheduler.engine.data.log.ErrorLogEvent;
import com.sos.scheduler.engine.kernel.util.ResourcePath;
import com.sos.scheduler.engine.test.binary.CppBinariesDebugMode;
import scala.Function1;
import scala.runtime.AbstractFunction1;

import javax.annotation.Nullable;

public final class TestConfigurationBuilder {
    private static final Function1<ErrorLogEvent, Boolean> defaultExpectedErrorLogEventPredicate = new AbstractFunction1<ErrorLogEvent, Boolean>() {
        @Override public Boolean apply(@Nullable ErrorLogEvent o) { return false; }
    };

    private final Class<?> testClass;
    private ResourcePath resourcePath;
    private Function1<ErrorLogEvent, Boolean> expectedErrorLogEventPredicate = defaultExpectedErrorLogEventPredicate;
    private CppBinariesDebugMode debugMode = CppBinariesDebugMode.debug;
    private String logCategories = System.getProperty("scheduler.logCategories");
    private boolean useDatabase = false;
    @Nullable private ImmutableMap<String,String> nameMap = null;
    @Nullable private ResourceToFileTransformer fileTransformer = null;

    public TestConfigurationBuilder(Class<?> testClass) {
        this.testClass = testClass;
        resourcePath = new ResourcePath(testClass.getPackage());
    }

    public TestConfigurationBuilder resourcesPackage(Package p) {
        resourcePath = new ResourcePath(p);
        return this;
    }

    public TestConfigurationBuilder nameMap(@Nullable ImmutableMap<String,String> o) {
        nameMap = o;
        return this;
    }

    public TestConfigurationBuilder resourceToFileTransformer(@Nullable ResourceToFileTransformer o) {
        fileTransformer = o;
        return this;
    }

    public TestConfigurationBuilder debugMode(CppBinariesDebugMode debugMode) {
        this.debugMode = debugMode;
        return this;
    }

    public TestConfigurationBuilder logCategories(String o) {
        this.logCategories = o;
        return this;
    }

    public TestConfigurationBuilder useDatabase(Boolean o) {
        this.useDatabase = o;
        return this;
    }

//    public TestConfiguration build() {
//        return new TestConfiguration(testClass, new Some(resourcePath), binariesDebugMode, logCategories,
//                useDatabase, expectedErrorLogEventPredicate,
//                nameMap, fileTransformer);
//    }
}

File: opensource/test/src/main/scala/com/sos/scheduler/engine/test/configuration/TestConfigurationBuilder.scala
package com.sos.scheduler.engine.test.configuration

import com.sos.scheduler.engine.test.ResourceToFileTransformer
import com.sos.scheduler.engine.test.binary.CppBinariesDebugMode
import scala.collection.JavaConversions._
import scala.collection.immutable

/** Für Java. */
final class TestConfigurationBuilder {
  private val default = TestConfiguration()
  private var _testPackage = default.testPackage
  private var _resourceToFileTransformer = default.resourceToFileTransformer
  private var _resourceNameMap = default.resourceNameMap
  private var _binariesDebugMode = default.binariesDebugMode
  private var _mainArguments = default.mainArguments
  private var _logCategories = default.logCategories
  private var _database = default.database
  private var _terminateOnError = default.terminateOnError

  def testPackage(o: Package) = {
    _testPackage = Some(o)
    this
  }

  def resourceToFileTransformer(o: ResourceToFileTransformer) = {
    _resourceToFileTransformer = Some(o)
    this
  }

  def resourceNameMap(o: java.util.Map[String, String]) = {
    _resourceNameMap = o.toMap
    this
  }

  def binariesDebugMode(o: CppBinariesDebugMode) = {
    _binariesDebugMode = Some(o)
    this
  }

  def mainArguments(o: java.lang.Iterable[String]) = {
    _mainArguments = immutable.Seq() ++ o
  }

  def logCategories(o: String) = {
    _logCategories = o
    this
  }

  def database(o: DatabaseConfiguration) = {
    _database = Some(o)
    this
  }

  def terminateOnError(o: Boolean) = {
    _terminateOnError = o
    this
  }

  def build: TestConfiguration = TestConfiguration(
    testPackage = _testPackage,
    resourceToFileTransformer = _resourceToFileTransformer,
    resourceNameMap = _resourceNameMap,
    binariesDebugMode = _binariesDebugMode,
    mainArguments = _mainArguments,
    logCategories = _logCategories,
    database = _database,
    terminateOnError= _terminateOnError
  )
}
